prog/hints.go

```diff
@@ -83,10 +123,182 @@ func generateHints(p *Prog, compMap CompMap, c *Call, arg Arg, exec func(newP *P
 	}
 }
 
-func checkConstArg(arg *ConstArg, compMap CompMap, cb func(newArg Arg)) {
-	for v, _ := range compMap[arg.Val] {
-		cb(constArg(arg.typ, v))
+// Creates hints for a ConstArg and calls cb() for each of the obtained
+// programs.
+func checkConstArg(arg *ConstArg, compMap CompMap, cb func(arg, newArg Arg)) {
+	replacersSet := getReplacersForVal(arg.Val, compMap)
+	for newV, _ := range replacersSet {
+		newArg := constArg(arg.typ, newV)
+		cb(arg, newArg)
+	}
+}
+
+// Returns a set of replacers for a given value. A replacer is a value with
+// which an argument's value should be replaced.
+func getReplacersForVal(value uint64, compMap CompMap) uint64Set {
+	replacersSet := make(uint64Set)
+	f := func(transform func(uint64) uint64) {
+		transformedValue := transform(value)
+		originalsSet := make(uint64Set)
+		// Get all different mutations of transformedValue: shrink, expand, ...
+		vals := getMutationsForConstVal(transformedValue)
+		diff := addArrayToSet(originalsSet, vals)
+		// Search for each unseen mutation.
+		for _, v := range diff {
+			compSet, ok := compMap[v]
+			if !ok {
+				continue
+			}
+			for newV, _ := range compSet {
+				// Transform the second operand the same way as the first one.
+				newV = transform(newV)
+				replacersSet[newV] = true
+			}
+		}
+	}
+	// Each transform is related to a different way of encoding an int.
+	transforms := []func(uint64) uint64{
+		// Identity transform.
+		func(v uint64) uint64 { return v },
+		// Revert order of bytes (Big Endian <-> Little Endian) and leave
+		// trailing zero bytes.
+		func(v uint64) uint64 { return reverse(v, false) },
```

Comment:
I think we need to deduplicate replacements across _all_ transformations. Otherwise we can get a bunch of duplicates.
For example, for a value 0xab, reverse(v, true) gives 0xab, so it's the same as no reverse (which we already checked).

Reply:
To answer myself: this is done by replacersSet.